internal/venice-common/src/main/java/com/linkedin/venice/utils/SharedObjectFactory.java

```diff
@@ -24,17 +32,23 @@ public class SharedObjectFactory<T> {
    * @return A shared object
    */
   public T get(String identifier, Supplier<T> constructor, Consumer<T> destroyer) {
-    if (!objectMap.containsKey(identifier)) {
-      T object = constructor.get();
-      objectMap.put(identifier, new ReferenceCounted<>(object, obj -> {
-        destroyer.accept(obj);
-        objectMap.remove(identifier);
-      }));
-      return object;
-    } else {
-      ReferenceCounted<T> referenceCounted = objectMap.get(identifier);
-      referenceCounted.retain();
+    Lock lock = getLock(identifier);
+    lock.lock();
+    try {
+      final AtomicBoolean newObjectCreated = new AtomicBoolean(false);
+      ReferenceCounted<T> referenceCounted = objectMap.computeIfAbsent(identifier, id -> {
+        newObjectCreated.set(true);
+        T object = constructor.get();
+        return new ReferenceCounted<>(object, destroyer);
+      });
+
+      if (!newObjectCreated.get()) {
```

Comment:
can we check the reference count of the object to find if its newly created or not?

Reply:
Yeah, I think it works now that I added the lock around this. Previously I hadn't added the lock and needed this. I prefer the solution that Gaojie suggested. If that doesn't work, I'll remove this `AtomicBoolean`

Reply:
With `compute`, I was able to get rid of this too